### ISSUE TITLE
Rspec fails with v0.6.6 when `eager_load` is set to `true`

### ISSUE THREAD

**Issue by gee-forr** [palkan/action_policy#252]
Hi there @palkan  and the rest of the martians :)

Thank you so much for this gem. It, and many more from Evil Martians make up a significant portion of our gemfiles. It is much appreciated.

I've got a weird bug that seems to have appeared with Action Policy v0.6.6 that goes away when I downgrade to v0.6.5. Below is more info.

### Tell us about your environment

**Ruby Version:** 3.2.2
**Framework Version:** Rails 7.0.8
**Action Policy Version:** 0.6.6
**Reproduction Script:** [action_policy_bug_repro.rb.txt](https://github.com/palkan/action_policy/files/12586173/action_policy_bug_repro.rb.txt) Please note that I couldnt get the repro script to trigger the error

### What did you do?

I run my tests with the following setting in `config/environments/test.rb`, and naturally, during CI, the $CI env var is set. This config helps speed up my test runs. In production, these values are hard-coded to `true`

```ruby
# somewhere in config/environments/test.rb
config.eager_load = ENV["CI"].present? # => true
```

I recently upgraded from v0.6.5 to v0.6.6 and now rspec runs fail immediately with the below error message. As soon as I unset the `CI` env var, and therefore disable eager loading, rspec runs perfectly.

If I downgrade to v0.6.5, everything works as normal with eager loading enabled.

### What actually happened?

RSpec fails immediately with the following error:

```
An error occurred while loading ./spec/actions/approval/gather_applicable_approvers_spec.rb.
Failure/Error: verify_authorized

NameError:
  undefined local variable or method `verify_authorized' for AuthorizedController:Class
# ./app/controllers/authorized_controller.rb:2:in `<class:AuthorizedController>'
# ./app/controllers/authorized_controller.rb:1:in `<top (required)>'
# /gems/ruby/3.2.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:30:in `require'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:135:in `const_get'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:135:in `cget'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:175:in `block in actual_eager_load_dir'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:40:in `block in ls'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:25:in `each'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:25:in `ls'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:170:in `actual_eager_load_dir'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:17:in `block (2 levels) in eager_load'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:16:in `each'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:16:in `block in eager_load'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:10:in `synchronize'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:10:in `eager_load'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader.rb:379:in `block in eager_load_all'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader.rb:377:in `each'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader.rb:377:in `eager_load_all'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/application/finisher.rb:74:in `block in <module:Finisher>'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:32:in `instance_exec'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:32:in `run'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:61:in `block in run_initializers'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:60:in `run_initializers'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/application.rb:372:in `initialize!'
# ./config/environment.rb:5:in `<top (required)>'
# ./spec/rails_helper.rb:5:in `<top (required)>'
# ./spec/actions/approval/gather_applicable_approvers_spec.rb:3:in `<top (required)>'
```

After that failure, every other spec immediately fails also with the following error message:

```
An error occurred while loading ./spec/actions/approval/gather_applicable_impact_approval_rules_spec.rb.
Failure/Error: require File.expand_path('../config/environment', __dir__)

FrozenError:
  can't modify frozen Array: [#<Pathname:/app/app/frontend/components>, "/app/app/actions", "/app/app/avo", "/app/app/channels", "/app/app/controllers", "/app/app/controllers/concerns", "/app/app/frontend", "/app/app/helpers", "/app/app/jobs", "/app/app/mailers", "/app/app/models", "/app/app/models/concerns", "/app/app/organizers", "/app/app/policies", "/app/app/processes", "/app/app/processes/concerns", "/app/app/serializers", "/app/app/serializers/concerns", "/app/app/snapshotters", "/app/app/values", "/gems/ruby/3.2.0/gems/sentry-rails-5.11.0/app/jobs", "/gems/ruby/3.2.0/gems/avo-2.40.0/app/components", "/gems/ruby/3.2.0/gems/avo-2.40.0/app/controllers", "/gems/ruby/3.2.0/gems/avo-2.40.0/app/helpers", "/gems/ruby/3.2.0/gems/view_component-3.5.0/app/controllers", "/gems/ruby/3.2.0/gems/view_component-3.5.0/app/controllers/concerns", "/gems/ruby/3.2.0/gems/view_component-3.5.0/app/helpers", "/gems/ruby/3.2.0/gems/lookbook-2.0.5/app/components", "/gems/ruby/3.2.0/gems/lookbook-2.0.5/app/controllers", "/gems/ruby/3.2.0/gems/lookbook-2.0.5/app/controllers/concerns", "/gems/ruby/3.2.0/gems/lookbook-2.0.5/app/helpers", "/gems/ruby/3.2.0/gems/actionmailbox-7.0.8/app/controllers", "/gems/ruby/3.2.0/gems/actionmailbox-7.0.8/app/jobs", "/gems/ruby/3.2.0/gems/actionmailbox-7.0.8/app/models", "/gems/ruby/3.2.0/gems/activestorage-7.0.8/app/controllers", "/gems/ruby/3.2.0/gems/activestorage-7.0.8/app/controllers/concerns", "/gems/ruby/3.2.0/gems/activestorage-7.0.8/app/jobs", "/gems/ruby/3.2.0/gems/activestorage-7.0.8/app/models", "test/components/previews", "/gems/ruby/3.2.0/gems/view_component-contrib-0.1.4/app/views"]
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/engine.rb:574:in `unshift'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/engine.rb:574:in `block in <class:Engine>'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:32:in `instance_exec'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:32:in `run'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:61:in `block in run_initializers'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:50:in `each'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:50:in `tsort_each_child'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:50:in `each'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:50:in `tsort_each_child'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/initializable.rb:60:in `run_initializers'
# /gems/ruby/3.2.0/gems/railties-7.0.8/lib/rails/application.rb:372:in `initialize!'
# ./config/environment.rb:5:in `<top (required)>'
# /gems/ruby/3.2.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:38:in `require'
# ./spec/rails_helper.rb:5:in `<top (required)>'
# /gems/ruby/3.2.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
# /gems/ruby/3.2.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:38:in `require'
# ./spec/actions/approval/gather_applicable_impact_approval_rules_spec.rb:3:in `<top (required)>'
```

**Comment by mdi** [comment]
I'm seeing the same issue as well.

**Comment by justinlittman** [comment]
I encountered a similar issue with v0.6.6 when deploying app to Heroku (but could not reproduce locally).

**Comment by gee-forr** [comment]
Updated the title and details to single out eager loading as the setting that triggers this issue.

The `cache_classes` setting is irrelevant in terms of this bug.

**Comment by gee-forr** [comment]
> I encountered a similar issue with v0.6.6 when deploying app to Heroku (but could not reproduce locally).

Try setting `config/environments/development.rb`'s `config.eager_load` to `true` and see if that reproduces the issue locally for you when trying to start your rails server?

**Comment by justinlittman** [comment]
Hmmm ... I couldn't reproduce with that setting change.

**Comment by t27duck** [comment]
This is a different error, but I believe is the same issue.

Here's a mostly minimal repo: https://github.com/t27duck/action_policy_eager_load

Check out, `bundle install`, and `rails c` should fail.

This is a vanilla rails app (minus action cable) with 0.6.6 installed with a single policy. The policy only has `relation_scope` in it. In `development.rb` the only change is setting `eager_load` to true.

Full stacktrace:

```
➜  action_policy_eager_load git:(main) rails c
/Users/tdrake/workbench/action_policy_eager_load/app/policies/some_policy.rb:2:in `<class:SomePolicy>': undefined method `relation_scope' for SomePolicy:Class (NoMethodError)
	from /Users/tdrake/workbench/action_policy_eager_load/app/policies/some_policy.rb:1:in `<main>'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:30:in `require'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:135:in `const_get'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:135:in `cget'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:175:in `block in actual_eager_load_dir'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:40:in `block in ls'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:25:in `each'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/helpers.rb:25:in `ls'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:170:in `actual_eager_load_dir'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:17:in `block (2 levels) in eager_load'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:16:in `each'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:16:in `block in eager_load'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:10:in `synchronize'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader/eager_load.rb:10:in `eager_load'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader.rb:379:in `block in eager_load_all'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader.rb:377:in `each'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/loader.rb:377:in `eager_load_all'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/application/finisher.rb:74:in `block in <module:Finisher>'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/initializable.rb:32:in `instance_exec'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/initializable.rb:32:in `run'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/initializable.rb:61:in `block in run_initializers'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:228:in `block in tsort_each'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:431:in `each_strongly_connected_component_from'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:349:in `block in each_strongly_connected_component'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:347:in `each'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:347:in `call'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:347:in `each_strongly_connected_component'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:226:in `tsort_each'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/3.0.0/tsort.rb:205:in `tsort_each'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/initializable.rb:60:in `run_initializers'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/application.rb:372:in `initialize!'
	from /Users/tdrake/workbench/action_policy_eager_load/config/environment.rb:5:in `<main>'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/zeitwerk-2.6.11/lib/zeitwerk/kernel.rb:38:in `require'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/application.rb:348:in `require_environment!'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/command/actions.rb:28:in `require_environment!'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/command/actions.rb:15:in `require_application_and_environment!'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/commands/console/console_command.rb:105:in `perform'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/thor-1.2.2/lib/thor/command.rb:27:in `run'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/thor-1.2.2/lib/thor/invocation.rb:127:in `invoke_command'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/thor-1.2.2/lib/thor.rb:392:in `dispatch'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/command/base.rb:87:in `perform'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/command.rb:48:in `invoke'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/railties-7.0.8/lib/rails/commands.rb:18:in `<main>'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from /Users/tdrake/.rbenv/versions/3.0.4/lib/ruby/gems/3.0.0/gems/bootsnap-1.16.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require'
	from bin/rails:4:in `<main>'
```

Setting `eager_load` to false does not invoke this exception.

**Comment by t27duck** [comment]
Also verified that 0.6.5 does not have this issue.

I have not disected anything, but my _guess_ is this one did it:

https://github.com/palkan/action_policy/commit/2a0531847d4a6aa0d8ac5f1639f4d0f415c6df8d

**Comment by palkan** [comment]
@t27duck Thanks a lot for the reproduction! Checking right now

**Comment by palkan** [comment]
Fixed in master. I'd appreciate If anyone has a chance to check the fix against the real app. In any case, will publish a new release in the end of the day.

**Comment by t27duck** [comment]
Verified `master` works with the minimal repo.

I switched `eager_load` to true in dev mode for a production app locally which fails to run `rails c` on 0.6.6 and switched the gem to `master`. It appears to boot properly now too.

**Comment by t27duck** [comment]
Just YOLO-ed a production app deploy that broke with 0.6.6 and pointed it to `master`. Restarts and runs just fine with this change.

**Comment by palkan** [comment]
Thanks!

Released in 0.6.7.

**Comment by gee-forr** [comment]
thanks so much @palkan and @t27duck and everyone else. Really appreciate the speedy fix.

@palkan - apologies that I wasn't able to effectively replicate the problem. Thank you again for all these amazing gems. Your work has made a materially positive difference to my team's ability to ship useful code in record time. ❤️
